pkgs/build-support/fetchgithub/default.nix

```
@@ -66,7 +68,22 @@ let
       inherit tag rev deepClone fetchSubmodules sparseCheckout fetchLFS; url = gitRepoUrl;
     } // lib.optionalAttrs (leaveDotGit != null) { inherit leaveDotGit; }
     else {
-      url = "${baseUrl}/archive/${revWithTag}.tar.gz";
+      # Use the API endpoint for private repos, as the archive URI doesn't
+      # support access with GitHub's fine-grained access tokens.
+      #
+      # Use the archive URI for non-private repos, as the API endpoint has
+      # relatively restrictive rate limits for unauthenticated users.
+      url =
+        let
+          endpoint = "/repos/${owner}/${repo}/tarball/${revWithTag}";
+        in
+        if private
+        then
+          if githubBase == "github.com"
+          then "https://api.github.com${endpoint}"
+          else "https://${githubBase}/api/v3${endpoint}"
+        else "${baseUrl}/archive/${revWithTag}.tar.gz";
```

Could scope it like this:

```suggestion
        if private then
          let
            endpoint = "/repos/${owner}/${repo}/tarball/${revWithTag}";
          in
          if githubBase == "github.com" then
            "https://api.github.com${endpoint}"
          else
            "https://${githubBase}/api/v3${endpoint}"
        else
          "${baseUrl}/archive/${revWithTag}.tar.gz";
```

What do you think? It's all the same in the end, just a narrower scope this way.

---

I think this is probably a bit clearer, yes! I'm not going to be able to make the change and update the PR until tomorrow, but I *think* you should be able to make the change yourself before merging if you want to get it done before then.

---

I feel the scope could be a bit more neater, but it looks good.